### ISSUE TITLE
HTTPAuthResource doesn't provide any way to support anonymous access.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2460]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#2460|
|Type|enhancement|
|Created|2007-02-19 19:02:14Z|

I think in the absence of an Authorization header portal.login should be called with twisted.cred.credentials.Anonymous, in the case where anonymous access is not allowed (i.e. no checker registered for the IAnonymous credentials interface) portal.login will errback and an unauthorized response will be sent.  Otherwise requestAvatar on the realm will be called normally.

This will also require a mechanism for resources below the HTTPAuthResource to trigger authentication at anypoint when the Anonymous credentials prove insufficient.  This could be a method on the IAuthenticatedRequest, perhaps IAuthenticatedRequest.reauthenticate()

<details><summary>Searchable metadata</summary>

```
trac-id__2460 2460
type__enhancement enhancement
reporter__dreid dreid
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1171911734000000 1171911734000000
changetime__1173121659000000 1173121659000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_1"></a>|@dreid set status to `assigned`|
|-|-|

[branches/web2-anon-auth-2460](branches/web2-anon-auth-2460)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_2"></a>|@dreid removed owner<br>@dreid set status to `new`|
|-|-|

This branch passes credentials.Anonymous to portal.login if it no header is given.  And adds a response filter that will add WWW-Authenticate headers to any UNAUTHORIZED responses that don't have them.  So triggering re authentication is as easy as raise HTTPError(401).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @dreid|
|-|-|

Looks very good. renderHTTP and locateChild miss docstrings in HTTPAuthResource, apart from that I think this is good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_5"></a>|@dreid set owner to @therve|
|-|-|

I added the docstring for renderHTTP and locateChild.  While writing the docstring for renderHTTP it dawned on me that renderHTTP's behavior was incorrect.  So I added a test for the proper behavior (it should call renderHTTP on the protected resource), and changed it's behavior to pass the test.  I don't think renderHTTP is ever meant to be called, but better safe than sorry.

-David

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @dreid|
|-|-|

Alright, but why locateChild didn't get the same treatment ? It seems it ends up the same, that's just to clarify it for me.

But it's a detail, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_7"></a>|@dreid commented|
|-|-|

To answer your question, the return value of locateChild is a resource and a sequence of path segments.  If segments isn't empty, the resource gets it's locateChild called.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid set status to `closed`|
|-|-|

(In [19764]) Merge web2-anon-auth-2460

Author: dreid
Reviewer: therve
Fixes [#2460](https://github.com/twisted/twisted/issues/2460)

Add support for anonymous authentication with the HTTPAuthResource.
HTTPAuthResource will attempt to authenticate with ANONYMOUS credentials
if no Authorization header is given.  If the portal has an AllowAnonymousAccess 
checker registered with it then this will succeed.  

If at any time an HTTPError is raised with an UNAUTHORIZED status code 
HTTPAuthResource will add WWW-Authenticate headers based on it's current
configuration (if the WWW-Authenticate header is absent)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
